### PR TITLE
Support adjustment of more CJK punctuation

### LINKED ISSUE #2439: CJK Punctuation Adjustment Not Comprehensive
### Description

I can see a lot have been done about punctuation adjustment since https://github.com/typst/typst/pull/954, but some punctuations are not covered, eg. question and exclamation marks.

image

Some common bracket variants are not adapted, either: `【`, `】`, `［`, `］`, `〖`, `〗`, `｛`, `｝`, `〔`, `〕`. I can't provide the complete list here. Maybe refer to the [Unicode Full-Width Forms](https://en.wikipedia.org/wiki/Halfwidth_and_Fullwidth_Forms_(Unicode_block)).

Also, [ctex](https://www.ctan.org/pkg/ctex) adopts five full-width punctuation compression schemes:

- `quanjiao` – Individual full-width punctuations take up `1em` in width, with two neighbouting ones taking up `1.5em` altogether.
- `banjiao` – All CJK punctuations take up `0.5em`.
- `kaiming` – Question/exclamation marks and full stops take up `1em`, all others `0.5em`.
- `CCT` – All punctuations are 'slightly narrower' than `1em` in width.
- `plain` – No adjustment.

Moreover, sources provided in https://github.com/typst/typst/issues/276 contain more specified guidelines on this issue. Thanks for the attention!

### Reproduction URL

https://typst.app/project/r0j15-4jgKGstUYllfWU9V

### Operating system

macOS

### Typst version

- [ ] I am using the latest version of Typst

### PR DESCRIPTION
**Pull request by Enter-tainer**
This PR add more(hopefully, all) CJK punctuations that should be adjusted. Specifically:

1. parentheses-like : `'】' | '〗' | '〕' | '〉' | '］' | '｝'`. see full list in https://www.w3.org/TR/clreq/#table_of_bracket_indication_punctuation_marks
2. full width stop mark `．`
3. `？` and `！` should be adjusted in GB style. see https://www.w3.org/TR/clreq/#h-punctuation_adjustment_space

A more comprehesive test which contains extensive usage of punctuations is also added

fix #2439

### CODE REVIEW

**Comment by Enter-tainer** [comment]
cc @peng1999 PTAL :heart:

File: tests/typ/layout/cjk-punctuation-adjustment.typ [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#set page(width: 15em)
+
+// In the following example, the space between 》 and ？ should be squeezed.
+// because zh-CN follows GB style
+#set text(lang: "zh", region: "CN", font: "Noto Serif CJK SC")
+原来，你也玩《原神》？
+
+// However, in the following example, the space between 》 and ？ should not be squeezed.
+// because zh-TW does not follow GB style
+#set text(lang: "zh", region: "TW", font: "Noto Serif CJK SC")
```
Comment (peng1999) [inline tests/typ/layout/cjk-punctuation-adjustment.typ:10]: It's better to use `Noto Serif CJK TC` here.

File: tests/ref/layout/cjk-punctuation-adjustment.png
Comment (peng1999) [inline tests/ref/layout/cjk-punctuation-adjustment.png:1]: `【` and `〔` at line start is not squeezed by 1/2 em. 

I guess the reason is that these punctuation marks are not considered to be able to appear at the beginning of a line. See `BEGIN_PUNCT_PAT` and `END_PUNCT_PAT` in `typst_library::layout::par::line` function for fix. Maybe the two consts should be moved to `shaping.rs` for contributors to recognize their relation more easily.

Comment (Enter-tainer) [reply tests/ref/layout/cjk-punctuation-adjustment.png:1]: Should I add `？` and `！` to `END_PUNCT_PAT`?

Comment (peng1999) [reply tests/ref/layout/cjk-punctuation-adjustment.png:1]: Yes.

**Comment by peng1999** [comment]
LGTM. Thanks for your work!

**Comment by laurmaedje** [comment]
Huge thanks for both of you working on this!

**Pull request merged**